File: src/similarity.c
/* $ID: similarity.c, last updated 2019/07/31, F.Osorio */

#include "similarity.h"
#include "spatialpack.h"
#include "stats.h"

void
SSIM(double *x, double *y, int *nr, int *nc, double *pars, double *eps, double *stats, double *comp)
{ /* compute structural similarity index for two images */
  int nobs, nrow = *nr, ncol = *nc;
  double xbar, ybar, xvar, yvar, cov;
  double c1, c2, c3;
  double alpha, beta, gamma;
  double luminance, contrast, structure, ssim;

  /* get constants and parameters */
  nobs = nrow * ncol;
  c1   = eps[0];
  c2   = eps[1];
  c3   = eps[2];
  alpha = pars[0];
  beta  = pars[1];
  gamma = pars[2];

  online_covariance(x, y, nobs, &xbar, &ybar, &xvar, &yvar, &cov);

  luminance = (2. * xbar * ybar + c1) / (SQR(xbar) + SQR(ybar) + c1);
  contrast  = (2. * sqrt(xvar) * sqrt(yvar) + c2) / (xvar + yvar + c2);
  structure = (cov + c3) / (sqrt(xvar) * sqrt(yvar) + c3);
  ssim = alpha * log(luminance) + beta * log(contrast) + gamma * log(structure);

  /* save results */
  stats[0] = xbar;
  stats[1] = ybar;
  stats[2] = xvar;
  stats[3] = yvar;
  stats[4] = cov;
  comp[0]  = exp(ssim);
  comp[1]  = luminance;
  comp[2]  = contrast;
  comp[3]  = structure;
}

void
CQ(double *x, double *y, int *nr, int *nc, int *h, double *eps, double *stats, double *comp)
{ /* compute CQ index for two images */
  int nobs, nrow = *nr, ncol = *nc;
  double xbar, ybar, xvar, yvar, cov;
  double c1, c2;
  double luminance, contrast, codispersion;

  /* get constants */
  nobs = nrow * ncol;
  c1   = eps[0];
  c2   = eps[1];

  online_covariance(x, y, nobs, &xbar, &ybar, &xvar, &yvar, &cov);

  luminance = (2. * xbar * ybar + c1) / (SQR(xbar) + SQR(ybar) + c1);
  contrast  = (2. * sqrt(xvar) * sqrt(yvar) + c2) / (xvar + yvar + c2);
  codisp_direction(x, y, &nrow, &ncol, h, comp); /* 1st element of 'comp' */
  codispersion = comp[0];

  /* save results */
  stats[0] = xbar;
  stats[1] = ybar;
  stats[2] = xvar;
  stats[3] = yvar;
  stats[4] = cov;
  comp[0]  = luminance * contrast * codispersion;
  comp[1]  = luminance;
  comp[2]  = contrast;
  comp[3]  = codispersion;
}

File: R/similarity.R
## $ID: similarity.R, last updated 2020/06/15, F.Osorio

SSIM <- function(x, y, alpha = 1, beta = 1, gamma = 1, eps = c(0.01, 0.03), L = 255)
{ ## structural similarity index for images (SSIM)
  # rescaling constants
  eps <- c(eps, 0)
  eps[1] <- (L * eps[1])^2
  eps[2] <- (L * eps[2])^2
  eps[3] <- .5 * eps[2]

  # coefficients of SSIM
  pars <- c(alpha, beta, gamma)

  if (all(dim(x)) != all(dim(y)))
    stop("x and y images must have same dimensions.")

  xrow <- nrow(x)
  xcol <- ncol(x)
  now  <- proc.time()

  # calling C function
  z <- .C("SSIM_coef",
          x = as.double(x),
          y = as.double(y),
          nrow = as.integer(xrow),
          ncol = as.integer(xcol),
          pars = as.double(pars),
          eps = as.double(eps),
          stats = double(5),
          comps = double(4))
  stats <- z$stats[c(1,3,2,4,5)]
  comps <- z$comps[2:4]
  ssim  <- z$comps[1]

  # output object
  speed <- proc.time() - now
  names(pars)  <- c("alpha", "beta", "gamma")
  names(stats) <- c("x.bar", "x.var", "y.bar", "y.var", "cov")
  names(comps) <- c("luminance", "contrast", "structure")
  o <- list(SSIM = ssim, coefficients = pars, comps = comps, stats = stats,
            speed = speed)
  o
}

CQ <- function(x, y, h = c(0,1), eps = c(0.01, 0.03), L = 255)
{ ## structural similarity index for images (CQ)
  # rescaling constants
  eps[1] <- (L * eps[1])^2
  eps[2] <- (L * eps[2])^2

  if (all(dim(x)) != all(dim(y)))
    stop("x and y images must have same dimensions.")

  if (!is.vector(h))
    stop("h must be a vector of directions.")
  if (length(h) > 2)
    warning("only the first two elements of h are used.")
  ok <- all(h >= 0)
  if (!ok)
    stop("only implemented for h >= 0.")

  xrow <- nrow(x)
  xcol <- ncol(x)
  now  <- proc.time()

  # calling C function
  z <- .C("CQ_coef",
          x = as.double(x),
          y = as.double(y),
          nrow = as.integer(xrow),
          ncol = as.integer(xcol),
          h = as.integer(h),
          eps = as.double(eps),
          stats = double(5),
          comps = double(4))
  stats <- z$stats[c(1,3,2,4,5)]
  comps <- z$comps[2:4]
  cq    <- z$comps[1]

  # output object
  speed <- proc.time() - now
  names(stats) <- c("x.bar", "x.var", "y.bar", "y.var", "cov")
  names(comps) <- c("luminance", "contrast", "structure")
  o <- list(CQ = cq, direction = h, comps = comps, stats = stats, speed = speed)
  o
}
